JavaScript/chapter01/p08_zero_matrix/avc278.js

```diff
@@ -0,0 +1,94 @@
+// Zero Matrix: Write an algorithm such that if an element in an MxN matrix is 0, its entire row and columns are set to 0.
+
+const assert = require("assert");
+
+/**
+ * Overwrite rows and columns who have at least 1 zero in them with zeros.
+ * @param  {Array.<number[]>} matrix input 2d array to mutate
+ * @return {null}
+ *
+ * Okay, so I have a jank solution that assumes no element in the array is currently -999.
+ * We iterate through the array, and when we find a 0, we flag its row and column by updating all elements in there to
+ * -999. The outer two for loop iterations is O(M * N) time, and each inner for loop is O(M) and O(N), respectively,
+ * resulting in a grand total of O((M * N) * (M + N)) time complexity, while updating takes O(1) additional space.
+ * When we take a second pass at the matrix, we update all elements that have been flagged -999 to 0, which is O(N^2)
+ * time and O(1) additional space.
+ *
+ * Runtime: O((M * N) * (M + N))
+ * Space:   O(1)
```

Comment:
can you think of a runtime/space tradeoff that can run in O(M * N) time?

Reply:
Yeah, the only other option I can think of would be to make an additional two arrays of bools O(M + N) set to false. Each time i, j in the original array is found to be zero, add i to one array, and j to the other. Once we're done iterating through the entire matrix, we can iterate through the first generated array of row bools and update all matrix elements in that row to 0s, then do the same to the column bools.
Runtime: O(M*N)
Space: O(M + N)

Reply:
yup, that works!